File: Training_Content/labs/lab03-migration-workflow/python/state.py
"""Migration Workflow Agent - State Management."""
from dataclasses import dataclass, field
from typing import List, Dict, Optional, Any
from enum import Enum


class Phase(Enum):
    """Migration phases."""
    ANALYSIS = "analysis"
    PLANNING = "planning"
    EXECUTION = "execution"
    VERIFICATION = "verification"
    COMPLETE = "complete"


@dataclass
class MigrationStep:
    """Represents a single migration step."""
    id: int
    description: str
    status: str = "pending"  # pending, in_progress, completed, failed
    input_files: List[str] = field(default_factory=list)
    output_files: List[str] = field(default_factory=list)
    result: Optional[str] = None


@dataclass
class MigrationState:
    """State for the migration workflow."""
    source_framework: str
    target_framework: str
    source_files: Dict[str, str]  # filename -> content
    phase: Phase = Phase.ANALYSIS
    analysis: Optional[Dict[str, Any]] = None
    plan: List[MigrationStep] = field(default_factory=list)
    current_step: int = 0
    migrated_files: Dict[str, str] = field(default_factory=dict)
    verification_result: Optional[Dict] = None
    errors: List[str] = field(default_factory=list)

File: Training_Content/labs/lab03-migration-workflow/typescript/src/state.ts
/**
 * Migration Workflow Agent - State Management
 */

import type {
  MigrationState,
  MigrationStep,
  Phase,
  AnalysisResult,
  VerificationResult,
} from './types.js';

/**
 * Create initial migration state
 */
export function createInitialState(
  sourceFramework: string,
  targetFramework: string,
  sourceFiles: Record<string, string>
): MigrationState {
  return {
    sourceFramework,
    targetFramework,
    sourceFiles,
    phase: 'analysis',
    analysis: null,
    plan: [],
    currentStep: 0,
    migratedFiles: {},
    verificationResult: null,
    errors: [],
  };
}

/**
 * Create a migration step
 */
export function createStep(
  id: number,
  description: string,
  inputFiles: string[] = []
): MigrationStep {
  return {
    id,
    description,
    status: 'pending',
    inputFiles,
    outputFiles: [],
    result: undefined,
  };
}

/**
 * Update state phase
 */
export function setPhase(state: MigrationState, phase: Phase): MigrationState {
  return { ...state, phase };
}

/**
 * Set analysis result
 */
export function setAnalysis(
  state: MigrationState,
  analysis: Record<string, AnalysisResult>
): MigrationState {
  return { ...state, analysis, phase: 'planning' };
}

/**
 * Set migration plan
 */
export function setPlan(
  state: MigrationState,
  plan: MigrationStep[]
): MigrationState {
  return { ...state, plan, phase: 'execution' };
}

/**
 * Update step status
 */
export function updateStepStatus(
  state: MigrationState,
  stepId: number,
  status: MigrationStep['status'],
  result?: string
): MigrationState {
  const plan = state.plan.map((step) =>
    step.id === stepId ? { ...step, status, result } : step
  );
  return { ...state, plan };
}

/**
 * Add migrated file
 */
export function addMigratedFile(
  state: MigrationState,
  filename: string,
  content: string
): MigrationState {
  return {
    ...state,
    migratedFiles: { ...state.migratedFiles, [filename]: content },
  };
}

/**
 * Increment current step
 */
export function incrementStep(state: MigrationState): MigrationState {
  return { ...state, currentStep: state.currentStep + 1 };
}

/**
 * Set verification result
 */
export function setVerificationResult(
  state: MigrationState,
  result: VerificationResult
): MigrationState {
  return { ...state, verificationResult: result, phase: 'complete' };
}

/**
 * Add error
 */
export function addError(state: MigrationState, error: string): MigrationState {
  return { ...state, errors: [...state.errors, error] };
}
